src/Storage/Notmuch.hs

```diff
@@ -1,39 +1,86 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
 -- | module for integrating notmuch within purebred
 module Storage.Notmuch where
 
 import Types (NotmuchMail(..))
 import Notmuch
 import Notmuch.Search
 
-import           Data.Maybe     (fromMaybe)
-import qualified Data.Vector    as Vec
-import           System.Process (readProcess)
+import Control.Monad (void)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Except (runExceptT)
+import Data.Traversable (traverse)
+import Data.List (union)
+import Data.Maybe (fromMaybe)
+import qualified Data.Vector as Vec
+import System.Process (readProcess)
+import Control.Exception (bracket)
 import qualified Data.Text as T
-import Data.Text.Encoding (decodeUtf8)
-import Types (NotmuchSettings, nmDatabase, mailTags)
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Types (NotmuchSettings, nmDatabase, mailId, mailTags)
 import Control.Lens.Getter (view)
+import Control.Lens.Setter (set)
 
 
 -- | creates a vector of parsed mails from a not much search
 -- Note, that at this point in time only free form searches are supported. Also,
 -- we filter out the tag which we use to mark mails as new mails
-getMessages :: T.Text -> NotmuchSettings FilePath -> IO (Vec.Vector NotmuchMail)
-getMessages s settings = do
-  db' <- databaseOpen (view nmDatabase settings)
-  case db' of
-    Left status -> do
-        error $ show status
-    Right db -> do
-        q <- query db (FreeForm $ T.unpack s)
-        msgs <- messages q
-        mails <- mapM messageToMail msgs
-        return $ Vec.fromList mails
+getMessages :: T.Text
+            -> NotmuchSettings FilePath
+            -> IO (Either String (Vec.Vector NotmuchMail))
+getMessages s settings = runExceptT
+  (do msgs <- databaseOpenReadOnly (view nmDatabase settings)
+              >>= flip query (FreeForm $ T.unpack s)
+              >>= messages
+      mails <- liftIO $ mapM messageToMail msgs
+      return $ Vec.fromList mails
+  ) >>= pure . either (Left . show) pure
```

Comment:
instead of
```haskell
runExceptT (k) >>= pure . either (Left . show) pure
```

Why not:
```haskell
first show <$> runExceptT (k)
```

(`Data.Bifunctor.first`)

Reply:
Also, does `getMessages` run multiple times?  Either we should open the database somewhere else and pass it around, or you should use `databaseDestroy` to close the database after each use (otherwise we leak memory
and probably other resources).

Reply:
Many thanks for the review. Yes `getMessages` is used a few times, certainly every time when you changes the search. Would it make sense to open the database with the config?

Reply:
One other thing which comes to my mind is using (resourceT)[https://hackage.haskell.org/package/resourcet] on the client side, to always close the database ... Currently I'd be leaning towards the resourcet approach and always closing the database after reading.

Reply:
ResourceT, maybe.  Let's not do it in this rebase.  Something to look at after perhaps.